controllers/discovery/service_controller.go

```diff
@@ -0,0 +1,245 @@
+				Namespace: service.Namespace,
+				Name:      service.Name,
+				Path:      nil,
+			},
+		},
+	}
+
+	// Let's find out the port, this annotation is a little bit more tricky.
+	if port, ok = service.Annotations["discovery.kuadrant.io/port"]; ok {
+		// check if the port is a number already.
+		if num, err := strconv.Atoi(port); err == nil {
+			int32num := int32(num)
+			Tag.Destination.Port = &int32num
+		} else {
+			// As the port is name, resolv the port from the service
+			for _, p := range service.Spec.Ports {
+				if p.Name == port {
+					Tag.Destination.Port = &p.Port
+					break
+				}
+			}
+		}
+	} else {
+		// As the annotation has not been set, let's check if the service has only on port, if that's the case,
+		//default to it.
+		if len(service.Spec.Ports) == 1 {
+			Tag.Destination.Port = &service.Spec.Ports[0].Port
+		}
+	}
+	// If we reach this point and the Port is still nil, this means bad news
+	if Tag.Destination.Port == nil {
+		return nil, fmt.Errorf("discovery.kuadrant.io/port is missing or invalid")
+	}
+
+	// Ok, let's get the configmap referenced by the annotation.
+	oasConfigmap := corev1.ConfigMap{}
+	err := r.Client.Get(ctx, types.NamespacedName{
+		Namespace: service.Namespace,
+		Name:      oasConfigmapName,
+	}, &oasConfigmap)
+	if err != nil {
+		return nil, err
+	}
+	// TODO(jmprusi): The openapispec.yaml data entry in the configmap is hardcoded, review this.
+	if _, ok := oasConfigmap.Data["openapi.yaml"]; !ok {
+		return nil, fmt.Errorf("oas configmap is missing the openapispec.yaml entry")
+	}
+	// TODO(jmprusi): We will create the API object in the same namespace as the service to simplify the deletion,
+	// review this later.
+	desiredAPI := v1beta1.API{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      apiName,
+			Namespace: service.Namespace,
+		},
+		Spec: v1beta1.APISpec{
+			APIDefinition: v1beta1.APIDefinition{
+				OAS: oasConfigmap.Data["openapi.yaml"],
+			},
+			TAGs: nil,
+		},
+	}
+
+	desiredAPI.Spec.TAGs = append(desiredAPI.Spec.TAGs, Tag)
+	return &desiredAPI, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		WithEventFilter(onlyLabeledServices()).
+		For(&corev1.Service{}).
+		Complete(r)
+}
+
+func onlyLabeledServices() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			// Lets filter for only Services that have the kuadrant label and are enabled.
+			if val, ok := e.Object.GetLabels()[kuadrantDiscoveryLabel]; ok {
+				enabled, _ := strconv.ParseBool(val)
+				return enabled
+			}
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// In case the update object had the kuadrant label set to true, we need to reconcile it.
+			if val, ok := e.ObjectOld.GetLabels()[kuadrantDiscoveryLabel]; ok {
+				enabled, _ := strconv.ParseBool(val)
+				return enabled
```

Comment:
If I remove the discovery label, the event is accepted. Should not be discarded only based on the new state?

Reply:
If you remove the `discovery label`, it means that the `discovery label` existed before, we need to handle that situation somehow, for example, removing the API object created by that service.